packages/block-editor/src/hooks/color.js

```diff
@@ -0,0 +1,247 @@
+/**
+ * External dependencies
+ */
+import { assign, has } from 'lodash';
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { __ } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+// This should be moved out of the components folder
+import {
+	getColorClassName,
+	getColorObjectByColorValue,
+	getColorObjectByAttributeValues,
+} from '../components/colors';
+import PanelColorSettings from '../components/panel-color-settings';
+import ContrastChecker from '../components/contrast-checker';
+import InspectorControls from '../components/inspector-controls';
+import { getBlockDOMNode } from '../utils/dom';
+
+export const COLOR_SUPPORT_KEY = '__experimentalColor';
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `backgroundColor` and `textColor` attribute.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// allow blocks to specify their own attribute definition with default values if needed.
+	if ( ! has( settings.attributes, [ 'backgroundColor', 'type' ] ) ) {
+		settings.attributes = assign( settings.attributes, {
+			backgroundColor: {
+				type: 'object',
+			},
+		} );
+	}
+	if ( ! has( settings.attributes, [ 'textColor', 'type' ] ) ) {
+		settings.attributes = assign( settings.attributes, {
+			textColor: {
+				type: 'object',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Override props assigned to save component to inject colors classnames.
+ *
+ * @param  {Object} props      Additional props applied to save element
+ * @param  {Object} blockType  Block type
+ * @param  {Object} attributes Block attributes
+ * @return {Object}            Filtered props applied to save element
+ */
+export function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, COLOR_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	// I'd have prefered to avoid the "style" attribute usage here
+	const { backgroundColor, textColor, style } = attributes;
+
+	const backgroundClass = getColorClassName(
+		'background-color',
+		backgroundColor
+	);
+	const textClass = getColorClassName( 'color', textColor );
+	props.className = classnames( props.className, backgroundClass, textClass, {
+		'has-text-color': textColor || style?.textColor,
+		'has-background': backgroundColor || style?.backgroundColor,
```

Comment:
The addition of these classes goes against a requirement @mtias referred: themes should be able to make a certain variable have a default in a given block and that default should apply to blocks that already existed.
In order to respect this requirement attributes part of "style" can not in any way change the markup of block besides setting CSS variables.

If a theme adds a background color to a block in blocks that already existed we will not have has-background as part of their markup.
Ideally, if CSS offered a way to select elements when a CSS variable is defined we would not need these classes anymore, but unfortunately, that does not seems possible.
A possible alternative would be to dynamically add these classes on the server.

Reply:
has-text-color is probably useless anyway and has-backrgound adds a padding but if we have padding controls in the block itself, it will conflict, so I guess it's also not that useful in the end. We might need to keep them for some time as BC and keep the styles for existing blocks.

Reply:
I think the auto-padding when setting a background color is pretty handy; padding controls on a block could just override it when set. Is it possible to keep this behavior somehow? It's kind of like setting a dark background color and automatically getting a light text color; it's a nice case of things "just working" by default.

Reply:
>  it possible to keep this behavior somehow?

Potentially, just set the padding value if we have a padding control. but we're not there, this PR won't change the current behavior.

Reply:
Yes, the "magic" behaviour can be nice, but also frustrating if you want flushed text with a background. Ideally, once we have padding control, it's just a matter of setting it by default on the same action, same as with text color.